experimental/devices/mfrc522/rf522.go

```diff
@@ -0,0 +1,763 @@
+	IrqPin        gpio.PinIO
+	Authenticated bool
+	antennaGain   int
+	MaxSpeedHz    int
+	spiDev        spi.Conn
+}
+
+// DefaultKey provides the default bytes for card authentication for method B.
+var DefaultKey = []byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+
+// NewSPI creates and initializes the RFID card reader attached to SPI.
+// - spiPort - the SPI device to use.
+// - maxSpeed - maximum speed for SPI data transfers.
+// - resetPin - reset GPIO pin.
+// - irqPin - irq GPIO pin.
+func NewSPI(spiPort spi.Port, maxSpeed int, resetPin, irqPin string) (device *Dev, err error) {
+
+	spiDev, err := spiPort.Connect(int64(maxSpeed), spi.Mode0, 8)
+	if err != nil {
+		return
+	}
+
+	if err != nil {
+		return
+	}
+
+	dev := &Dev{
+		spiDev:     spiDev,
+		MaxSpeedHz: maxSpeed,
+	}
+
+	pin := gpioreg.ByName(resetPin)
+	dev.ResetPin = pin
+	dev.ResetPin.Out(gpio.High)
+
+	pin = gpioreg.ByName(irqPin)
+	dev.IrqPin = pin
+	dev.IrqPin.In(gpio.PullUp, gpio.FallingEdge)
+
+	err = dev.Init()
+
+	device = dev
+
+	return
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() (err error) {
+	err = r.Reset()
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TModeReg, 0x8D)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TPrescalerReg, 0x3E)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegL, 30)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegH, 0)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TxAutoReg, 0x40)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.ModeReg, 0x3D)
+	if err != nil {
+		return
+	}
+	err = r.SetAntenna(true)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (r *Dev) writeSpiData(dataIn []byte) (out []byte, err error) {
+	out = make([]byte, len(dataIn))
+	err = r.spiDev.Tx(dataIn, out)
+	return
+}
+
+func printBytes(data []byte) (res string) {
+	res = "["
+	for _, v := range data[0 : len(data)-1] {
+		res = res + fmt.Sprintf("%02x, ", byte(v))
+	}
+	res = res + fmt.Sprintf("%02x", data[len(data)-1])
+	res = res + "]"
+	return
+}
+
+func (r *Dev) devWrite(address int, data byte) (err error) {
+	newData := [2]byte{(byte(address) << 1) & 0x7E, data}
+	_, err = r.writeSpiData(newData[:])
+	return
+}
+
+func (r *Dev) devRead(address int) (result byte, err error) {
+	data := [2]byte{((byte(address) << 1) & 0x7E) | 0x80, 0}
+	rb, err := r.writeSpiData(data[:])
+	result = rb[1]
+	return
+}
+
+func (r *Dev) setBitmask(address, mask int) (err error) {
+	current, err := r.devRead(address)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(address, current|byte(mask))
+	return
+}
+
+func (r *Dev) clearBitmask(address, mask int) (err error) {
+	current, err := r.devRead(address)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(address, current&^byte(mask))
+	return
+
+}
+
+// SetAntennaGain initiates the antenna gain, values must be between 0 and 7.
+func (r *Dev) SetAntennaGain(gain int) (err error) {
+	if 0 <= gain && gain <= 7 {
```

Comment:
```
if gain < 0 || gain > 7 {
  return errors.New("mfrc522: invalid gain value")
}
r.antennaGain = gain
return
```

But I'm concerned that it doesn't send the command to the device, that's confusing.